core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -226,10 +241,10 @@ public boolean isCollectingTaskNeeded(QueryId queryId, PlanFragment plan)
      * In such case dynamic filters must be unblocked (and probe split generation resumed) for
      * source stage containing joins to allow build source tasks to flush data and complete.
      */
-    public void unblockStageDynamicFilters(QueryId queryId, PlanFragment plan)
+    public void unblockStageDynamicFilters(QueryId queryId, int attemptId, PlanFragment plan)
     {
         DynamicFilterContext context = dynamicFilterContexts.get(queryId);
-        if (context == null) {
+        if (context == null || attemptId != context.getCurrentAttemptId()) {
```

Comment:
I would make it explicit `attemptId < context.getCurrentAttemptId()`

Comment below needs update